service/config_provider.go

```diff
@@ -75,6 +75,37 @@ type configProvider struct {
 	watcher chan error
 }
 
+// ConfigProviderOptions has options that change the behavior of ConfigProvider
+// returned by NewConfigProvider()
+type ConfigProviderOptions struct {
+	configMapProviders map[string]configmapprovider.Provider
+	cfgMapConverters   []config.MapConverterFunc
+	configUnmarshaler  configunmarshaler.ConfigUnmarshaler
+}
+
+// ConfigProviderOption is an option to change the behavior of ConfigProvider
+// returned by NewConfigProvider()
+type ConfigProviderOption func(opts *ConfigProviderOptions)
+
+func WithConfigMapProviders(c map[string]configmapprovider.Provider) ConfigProviderOption {
+	return func(opts *ConfigProviderOptions) {
+		opts.configMapProviders = c
+	}
+}
```

Comment:
Not sure if we want a map as argument or one of the next two options:

1.
```golang
func WithConfigMapProvider(scheme string, p configmapprovider.Provider) ConfigProviderOption {
	return func(opts *ConfigProviderOptions) {
		opts.configMapProviders[scheme] = p
	}
}
```

2.
```golang

// Change configmapprovider.Provider to have a `Scheme()` func on it.

func WithConfigMapProvider(p configmapprovider.Provider) ConfigProviderOption {
	return func(opts *ConfigProviderOptions) {
		opts.configMapProviders[p.Scheme()] = p
	}
}
```

Reply:
I wanted to implement both approaches in seperate commits but I have an issue with the first.

Due to the fact that 2 of the functions (`service.MustNewConfigProvider` and `service.TestConfigProvider_Errors`) that use `WithConfigMapProviders` are using `map[string]configmapprovider.Provider` approach I need to translate these maps to `[]ConfigProviderOption` and after that append this array to other Options and pass them to `NewConfigProvider` function.

example:
```go
// Convert map[string]configmapprovider.Provider to []ConfigProviderOption

configMapProvidersOpts := []ConfigProviderOption{}
for scheme, provider := range configMapProviders {
configMapProvidersOpts = append(configMapProvidersOpts, WithConfigMapProvider(scheme, provider))
}

configProvider, err := NewConfigProvider(locations, append(configMapProvidersOpts, WithCfgMapConverters(cfgMapConverters), WithConfigUnmarshaler(configUnmarshaler))...)
```

should I create a function that converts config maps to `[]ConfigProviderOption` and appends them to provided options 
to avoid reused code or it's not that big deal?

```go
func AppendConfigMapProvider(
	opts []ConfigProviderOption,
	configMapProviders map[string]configmapprovider.Provider) []ConfigProviderOption {
	// Convert map[string]configmapprovider.Provider to []ConfigProviderOption
	for scheme, provider := range configMapProviders {
		opts = append(opts, WithConfigMapProvider(scheme, provider))
	}
	return opts
}

```

Reply:
> // Change configmapprovider.Provider to have a `Scheme()` func on it.

I think I prefer this approach.  We require that providers have a scheme anyways, so why not a way to determine what they think their scheme is?

> should I create a function that converts config maps to []ConfigProviderOption and appends them to provided options to avoid reused code or it's not that big deal?

An unexported helper function to do this might be useful, but at least `service.MustNewConfigProvider()` is expected to go away in the near future, so I'm not sure how big a deal it is.

Reply:
> I think I prefer this approach. We require that providers have a scheme anyways, so why not a way to determine what they think their scheme is?

After more thoughts on this, I think is the right thing to do, since this will allow us to use the same pattern in the builder to support more providers - builder will look for NewProvider in the given import (similar to NewFactory for components.

Per @Aneurysm9 comment the `service.MustNewConfigProvider()` is going away, and the other one is a test. I would rather look into an example how a distribution with extra providers will use this. Or an example written for this case.